flocker/node/test/test_script.py

```diff
@@ -71,137 +81,325 @@ def deployer_factory_stub(**kw):
+                    needs_reactor=False, needs_id=True,
+                    api_factory=API, deployer_type=u"p2p",
+                ),
+            ],
+        )
+        api = agent_service.get_api()
+        self.assertEqual(
+            API(cluster_id=self.ca_set.node.cluster_uuid),
+            api,
+        )
 
-    def test_config_validated(self):
+
+class AgentServiceDeployerTests(SynchronousTestCase):
+    """
+    Tests for ``AgentService.get_deployer``.
+    """
+    setUp = agent_service_setup
+
+    def test_backend_selection(self):
         """
-        ``GenericAgentScript.main`` validates the configuration file.
+        ``AgentService.get_deployer`` finds a factory the configured backend in
+        its ``deployers`` dictionary and uses it to create the new
+        ``IDeployer`` provider with the supplied API object and the local
+        host's external IP address and unique node identifier.
         """
-        self.config.setContent("INVALID")
+        ip = b"192.0.2.7"
+        ips = {
+            (self.agent_service.control_service_host,
+             self.agent_service.control_service_port): ip,
+        }
 
-        service = Service()
-        options = DatasetAgentOptions()
-        options.parseOptions([b"--agent-config", self.config.path])
-        test_reactor = MemoryCoreReactor()
+        class Deployer(PRecord):
+            api = field(mandatory=True)
+            hostname = field(mandatory=True)
+            node_uuid = field(mandatory=True)
+
+        class WrongDeployer(PRecord):
+            pass
+
+        def get_external_ip(host, port):
+            return ips[host, port]
+
+        agent_service = self.agent_service.set(
+            "get_external_ip", get_external_ip,
+        ).set(
+            "backends", [
+                BackendDescription(
+                    name=self.agent_service.backend_name,
+                    needs_reactor=False, needs_id=False,
+                    api_factory=None, deployer_type=u"magic",
+                ),
+            ],
+        ).set(
+            "deployers", {
+                u"magic": Deployer,
+                b"bar": WrongDeployer,
+            },
+        )
 
-        self.assertRaises(
-            ValidationError,
-            GenericAgentScript().main, test_reactor, options, service,
+        api = object()
+        deployer = agent_service.get_deployer(api)
+        self.assertEqual(
+            Deployer(
+                api=api,
+                hostname=ip,
+                node_uuid=self.ca_set.node.uuid,
+            ),
+            deployer,
         )
 
-    def test_missing_configuration_file(self):
+
+class AgentServiceTLSContextTests(SynchronousTestCase):
+    """
+    Tests for ``AgentService.get_tls_context``.
+    """
+    setUp = agent_service_setup
+
+    def test_x(self):
+        # XXX Assertions are so hard :(
+        tls_context = self.agent_service.get_tls_context()
+        tls_context
```

Comment:
Well I won't try and figure this out. `_TLSContext` is a `PRecord` with attributes that should match the `test_case.ca_set` I guess.
:question: Either finish off or delete and create a followup issue.